### PR TITLE
[Feature] Support function last_day

### LINKED ISSUE #19548: [function] last_day
## Feature request

as mentioned in #15385, last_day need to be supported. However last day can be used in month/quarter/year. So I think it's better to support this feature flexibly.

here is my proposal:

### description
Returns the last day of the specified date part for a date or datetime.

### syntax
```
LAST_DAY ( <date_or_datetime_expr> , [<date_part>])
```
in which, 
 `date_or_datetime_expr` is a date or datetime expression
`date_part` is an optional parameter. can be `month`, `quarter`,`year`, default to `month`. 
the return value is always a date.

### PR DESCRIPTION
**Pull request by vendanner**
## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes:

Fixes #19548

## Problem Summary(Required):

NEW FUNCTION: `last_day`

Test sql:
```sql
CREATE TABLE IF NOT EXISTS `last_day_table` (
  `d1` DATE,
  `d2` DATETIME
)
DISTRIBUTED BY HASH(`d1`)
PROPERTIES(
  "replication_num" = "1"
);

INSERT INTO `last_day_table`(d1, d2)
VALUES ('2020-02-12', '2020-02-12 08:08:14'),
       ('2021-03-28', '2021-03-28 08:08:14'),
       ('2022-04-28', '2022-04-28 08:08:14'),
       ('2023-05-29', '2023-05-29 08:08:14');

CREATE TABLE IF NOT EXISTS `last_day_with_optional_table` (
  `d1` DATE,
  `d2` DATETIME,
  `optional` CHAR(8)
)
DISTRIBUTED BY HASH(`d1`)
PROPERTIES(
  "replication_num" = "1"
);

INSERT INTO `last_day_with_optional_table`(d1, d2, optional)
VALUES ('2020-02-12', '2020-02-12 08:08:14', 'month'),
       ('2021-03-28', '2021-03-28 08:08:14', 'quarter'),
       ('2022-04-28', '2022-04-28 08:08:14', 'year'),
       ('2023-05-29', '2023-05-29 08:08:14', 'year');

```
Test results:
```sql
mysql> select last_day('2023-02-12', 'month');
+---------------------------------+
| last_day('2023-02-12', 'month') |
+---------------------------------+
| 2023-02-28                      |
+---------------------------------+

mysql> select last_day('2020-02-12');
+------------------------+
| last_day('2020-02-12') |
+------------------------+
| 2020-02-29             |
+------------------------+

mysql> select last_day('2023-05-10', 'quarter');
+-----------------------------------+
| last_day('2023-05-10', 'quarter') |
+-----------------------------------+
| 2023-06-30                        |
+-----------------------------------+

mysql> select last_day('2023-05-10', 'year');
+--------------------------------+
| last_day('2023-05-10', 'year') |
+--------------------------------+
| 2023-12-31                     |
+--------------------------------+

mysql> select d1, last_day(d1, 'month'), d2, last_day(d2, 'month') from last_day_table order by d1;
+------------+-----------------------+---------------------+-----------------------+
| d1         | last_day(d1, 'month') | d2                  | last_day(d2, 'month') |
+------------+-----------------------+---------------------+-----------------------+
| 2020-02-12 | 2020-02-29            | 2020-02-12 08:08:14 | 2020-02-29            |
| 2021-03-28 | 2021-03-31            | 2021-03-28 08:08:14 | 2021-03-31            |
| 2022-04-28 | 2022-04-30            | 2022-04-28 08:08:14 | 2022-04-30            |
| 2023-05-29 | 2023-05-31            | 2023-05-29 08:08:14 | 2023-05-31            |
+------------+-----------------------+---------------------+-----------------------+

mysql> select d1, optional, last_day(d1, optional), d2, optional, last_day(d2, optional) from last_day_with_optional_table order by d1;
+------------+----------+------------------------+---------------------+----------+------------------------+
| d1         | optional | last_day(d1, optional) | d2                  | optional | last_day(d2, optional) |
+------------+----------+------------------------+---------------------+----------+------------------------+
| 2020-02-12 | month    | 2020-02-29             | 2020-02-12 08:08:14 | month    | 2020-02-29             |
| 2021-03-28 | quarter  | 2021-03-31             | 2021-03-28 08:08:14 | quarter  | 2021-03-31             |
| 2022-04-28 | year     | 2022-12-31             | 2022-04-28 08:08:14 | year     | 2022-12-31             |
| 2023-05-29 | year     | 2023-12-31             | 2023-05-29 08:08:14 | year     | 2023-12-31             |
+------------+----------+------------------------+---------------------+----------+------------------------+
```
## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [x] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by LiShuMing** [comment]
Thanks for your contributions!

can you fix the DCO error? 
```
There are 2 commits incorrectly signed off. This means that the author(s) of these commits failed to include a Signed-off-by line in their commit message.

To avoid having PRs blocked in the future, always include Signed-off-by: Author Name <authoremail@example.com> in every commit message. You can also do this automatically by using the -s flag (i.e., git commit -s).

Here is how to fix the problem so that this code can be merged.
```

Like other prs, can you add some tests in `test/sql/test_time_fn/R/test_time_fn` (eg: https://github.com/StarRocks/starrocks/pull/21135/files)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by LiShuMing — DISMISSED** [review]

File: be/src/exprs/time_functions.cpp [hunk 1]
```diff
@@ -2361,4 +2361,125 @@ StatusOr<ColumnPtr> TimeFunctions::make_date(FunctionContext* context, const Col
...
+    ColumnBuilder<TYPE_DATE> result(size);
+    for (int row = 0; row < size; ++row) {
+        if (data_column.is_null(row) || format_column.is_null(row)) {
+            result.append_null();
+            continue;
+        }
+
+        DateValue date = (DateValue)data_column.value(row);
+        Slice format_content = format_column.value(row);
+        if (!(format_content == "year" || format_content == "month" || format_content == "quarter")) {
```
Comment (mofeiatwork) [inline be/src/exprs/time_functions.cpp:2423]: compare ignore case ?

**Review by murphyatwork — CHANGES_REQUESTED** [review]

File: be/src/exprs/time_functions.cpp (hunk 1 above)
Comment (vendanner) [reply be/src/exprs/time_functions.cpp:2423]: https://github.com/StarRocks/starrocks/issues/19548
"date_part is an optional parameter. can be month, quarter,year, default to month. "  must be lower case

File: be/src/exprs/time_functions.cpp [hunk 2]
```diff
@@ -2361,4 +2361,125 @@ StatusOr<ColumnPtr> TimeFunctions::make_date(FunctionContext* context, const Col
...
+    context->set_function_state(scope, state);
+
+    if (!context->is_notnull_constant_column(1)) {
+        return Status::OK();
+    }
+
+    state->const_optional = true;
+    ColumnPtr column = context->get_constant_column(1);
+    Slice optional = ColumnHelper::get_const_value<TYPE_VARCHAR>(column);
+    if (!(optional == "year" || optional == "month" || optional == "quarter")) {
```
Comment (satanson) [inline be/src/exprs/time_functions.cpp:2465]: case-sensitive?  can we use Year, MoNth, quartER?

File: be/src/exprs/time_functions.cpp [hunk 3]
```diff
@@ -2361,4 +2361,125 @@ StatusOr<ColumnPtr> TimeFunctions::make_date(FunctionContext* context, const Col
...
+            date.set_end_of_quarter();
+        } else {
+            date.set_end_of_month();
+        }
+        result.append(date);
+    }
+    return result.build(ColumnHelper::is_all_const(columns));
+}
+
+StatusOr<ColumnPtr> TimeFunctions::_last_day_with_format_optional(FunctionContext* context, const Columns& columns) {
```
Comment (satanson) [inline be/src/exprs/time_functions.cpp:2409]: optional means something is present or absent,  here the format argument is absolutely present , so can we just use   name  "_last_day_with_format" to indicates it is a default and non-optimized branch of last_day?

File: be/src/exprs/time_functions.cpp (hunk 2 above)
Comment (vendanner) [reply be/src/exprs/time_functions.cpp:2465]: case-sensitive, only "year"、"month"、"quarter"

File: be/src/exprs/time_functions.cpp (hunk 3 above)
Comment (vendanner) [reply be/src/exprs/time_functions.cpp:2409]: ok, I will last_day_with_format_optional modify _last_day_with_format

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by satanson — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Comment by silverbullet233** [comment]
@mergify backport branch-3.0

**Pull request merged**